src/net/client/cache.rs

```diff
@@ -0,0 +1,1200 @@
+    }
+
+    /// Set the maximum time to cache NXDOMAIN results.
+    ///
+    /// The value has to be at least 60 seconds (one minute), at most 86,400
+    /// seconds (one day) and the default is 3,600 seconds (one hour).
+    pub fn set_max_nxdomain_validity(&mut self, value: Duration) {
+        self.max_nxdomain_validity = MAX_NXDOMAIN_VALIDITY.limit(value)
+    }
+
+    /// Set the maximum time to cache NODATA results.
+    ///
+    /// The value has to be at least 60 seconds (one minute), at most 86,400
+    /// seconds (one day) and the default is 3,600 seconds (one hour).
+    pub fn set_max_nodata_validity(&mut self, value: Duration) {
+        self.max_nodata_validity = MAX_NODATA_VALIDITY.limit(value)
+    }
+
+    /// Set the maximum time to cache delegations.
+    ///
+    /// The value has to be at least 60 seconds (one minute), at most
+    /// 1,000,000,000 seconds and the default is 1,000,000 seconds.
+    pub fn set_max_delegation_validity(&mut self, value: Duration) {
+        self.max_delegation_validity = MAX_DELEGATION_VALIDITY.limit(value)
+    }
+}
+
+impl Default for Config {
+    fn default() -> Self {
+        Self {
+            max_cache_entries: MAX_CACHE_ENTRIES.default(),
+            max_validity: MAX_VALIDITY.default(),
+            transport_failure_duration: TRANSPORT_FAILURE_DURATION.default(),
+            misc_error_duration: MISC_ERROR_DURATION.default(),
+            max_nxdomain_validity: MAX_NXDOMAIN_VALIDITY.default(),
+            max_nodata_validity: MAX_NODATA_VALIDITY.default(),
+            max_delegation_validity: MAX_DELEGATION_VALIDITY.default(),
+        }
+    }
+}
+
+//------------ Connection -----------------------------------------------------
+
+#[derive(Clone)]
+/// A connection that caches responses from an upstream connection.
+pub struct Connection<Upstream, C: Clock + Send + Sync = SystemClock> {
+    /// Upstream transport to use for requests.
+    upstream: Upstream,
+
+    /// The cache for this connection.
+    cache: Cache<Key, Arc<Value<C>>>,
+
+    /// The configuration of this connection.
+    config: Config,
+
+    /// The clock to use for expiring cache entries.
+    clock: C,
+}
+
+impl<Upstream> Connection<Upstream> {
+    /// Create a new connection with default configuration parameters.
+    pub fn new(upstream: Upstream) -> Self {
+        Self::with_config(upstream, Default::default())
+    }
+
+    /// Create a new connection with specified configuration parameters.
+    pub fn with_config(upstream: Upstream, config: Config) -> Self {
+        Self {
+            upstream,
+            cache: Cache::new(config.max_cache_entries),
+            config,
+            clock: SystemClock::new(),
+        }
+    }
+}
+
+impl<Upstream, C> Connection<Upstream, C>
+where
+    C: Clock + Send + Sync + 'static,
+{
+    /// Create a new connection with default configuration parameters.
+    pub fn new_with_time(upstream: Upstream, clock: C) -> Self {
```

Comment:
If kept this should be documented as being intended for internal test purposes only. Adopting https://github.com/NLnetLabs/domain/pull/278 however would remove this issue by removing these `with_time` functions.